drivers/storage/efs/storage/efs_storage.go

```diff
@@ -0,0 +1,592 @@
+// Volumes returns all volumes or a filtered list of volumes.
+func (d *driver) Volumes(
+	ctx types.Context,
+	opts *types.VolumesOpts) ([]*types.Volume, error) {
+
+	fileSystems, err := d.getAllFileSystems()
+	if err != nil {
+		return nil, err
+	}
+
+	var volumesSD []*types.Volume
+	for _, fileSystem := range fileSystems {
+		// Only volumes with partition prefix
+		if !strings.HasPrefix(*fileSystem.Name, d.tag()+tagDelimiter) {
+			continue
+		}
+
+		// Only volumes in "available" state
+		if *fileSystem.LifeCycleState != awsefs.LifeCycleStateAvailable {
+			continue
+		}
+
+		volumeSD := &types.Volume{
+			Name:        d.getPrintableName(*fileSystem.Name),
+			ID:          *fileSystem.FileSystemId,
+			Size:        *fileSystem.SizeInBytes.Value,
+			Attachments: nil,
+		}
+
+		var atts []*types.VolumeAttachment
+		if opts.Attachments {
+			atts, err = d.getVolumeAttachments(ctx, *fileSystem.FileSystemId)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if len(atts) > 0 {
+			volumeSD.Attachments = atts
+		}
+		volumesSD = append(volumesSD, volumeSD)
+	}
+
+	return volumesSD, nil
+}
+
+// VolumeInspect inspects a single volume.
+func (d *driver) VolumeInspect(
+	ctx types.Context,
+	volumeID string,
+	opts *types.VolumeInspectOpts) (*types.Volume, error) {
+
+	resp, err := d.efsClient().DescribeFileSystems(&awsefs.DescribeFileSystemsInput{
+		FileSystemId: aws.String(volumeID),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(resp.FileSystems) > 0 {
+		fileSystem := resp.FileSystems[0]
+
+		// Only volumes in "available" state
+		if *fileSystem.LifeCycleState != awsefs.LifeCycleStateAvailable {
+			return nil, nil
+		}
+
+		volume := &types.Volume{
+			Name:        d.getPrintableName(*fileSystem.Name),
+			ID:          *fileSystem.FileSystemId,
+			Size:        *fileSystem.SizeInBytes.Value,
+			Attachments: nil,
+		}
+
+		var atts []*types.VolumeAttachment
+
+		if opts.Attachments {
+			atts, err = d.getVolumeAttachments(ctx, *fileSystem.FileSystemId)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if len(atts) > 0 {
+			volume.Attachments = atts
+		}
+		return volume, nil
+	}
+
+	return nil, types.ErrNotFound{}
+}
+
+// VolumeCreate creates a new volume.
+func (d *driver) VolumeCreate(
+	ctx types.Context,
+	name string,
+	opts *types.VolumeCreateOpts) (*types.Volume, error) {
+
+	request := &awsefs.CreateFileSystemInput{
+		CreationToken:   aws.String(name),
+		PerformanceMode: aws.String(awsefs.PerformanceModeGeneralPurpose),
+	}
+	if opts.Type != nil && strings.ToLower(*opts.Type) == "maxio" {
+		request.PerformanceMode = aws.String(awsefs.PerformanceModeMaxIo)
+	}
+	fileSystem, err := d.efsClient().CreateFileSystem(request)
+
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = d.efsClient().CreateTags(&awsefs.CreateTagsInput{
+		FileSystemId: fileSystem.FileSystemId,
+		Tags: []*awsefs.Tag{
+			{
+				Key:   aws.String("Name"),
+				Value: aws.String(d.getFullVolumeName(name)),
+			},
+		},
+	})
+
+	if err != nil {
```

Comment:
should we attempt to purge the volume here if tag creation fails to avoid leaking volumes?